### PR TITLE
Staged replacement of the use of evaluate* and get* with evaluate() and value() [1/N] 

### PR DESCRIPTION
**Pull request by protonu**
This replaces uses of evaluate(Int, Bool, Double) with evaluate which returns a PolymorphicValue.

This also replaces the uses of get(Int, Bool, Double) with value().

This PR is the first of a number of them to fix the issue. In the final PR, the older get* and evaluate* methods will be removed.

Fixes https://github.com/NVIDIA/Fuser/issues/643

### CODE REVIEW

**Comment by protonu** [comment]
!build

File: csrc/ir/base_nodes.cpp [hunk 1]
```diff
@@ -291,30 +291,44 @@ std::optional<bool> Val::getBool() const {
...
+      evaluated_val.hasValue(),
+      "Detected a const value but failed to infer its value: ",
+      toInlineString());
+  return evaluated_val;
+}
+
 bool Val::isZero() const {
-  return getInt() == 0 || getDouble() == 0.0;
+  return value().hasValue() &&
+      (value().is<double>() || value().is<int64_t>()) && value() == 0;
```
Comment (zasdfgbnm) [inline csrc/ir/base_nodes.cpp:310]:
```suggestion
  return value().hasValue() && value() == 0;
```

File: csrc/ir/base_nodes.cpp [hunk 2]
```diff
@@ -291,30 +291,44 @@ std::optional<bool> Val::getBool() const {
...
 bool Val::isZeroInt() const {
-  auto int_val = getInt();
-  return int_val.has_value() && int_val.value() == 0;
+  return value().hasValue() && value().is<int64_t>() && value() == 0;
 }
 
 bool Val::isOne() const {
-  return getInt() == 1 || getDouble() == 1.0;
+  return value().hasValue() &&
+      (value().is<double>() || value().is<int64_t>()) && value() == 1;
```
Comment (zasdfgbnm) [inline csrc/ir/base_nodes.cpp:319]:
```suggestion
  return value().hasValue() && value() == 1;
```

File: test/test_optimization_pass.cpp [hunk 3]
```diff
@@ -655,7 +655,8 @@ TEST_F(NVFuserTest, FusionRemoveEmptyMatmul_CUDA) {
   EXPECT_NE(preseg_fusion->outputs()[0]->definition(), nullptr);
   auto rewritten_def = preseg_fusion->outputs()[0]->definition();
   EXPECT_TRUE(rewritten_def->isA<FullOp>());
-  EXPECT_EQ(rewritten_def->as<FullOp>()->getFillValue()->evaluateDouble(), 0.0);
+  ExpressionEvaluator ee;
```
Comment (zasdfgbnm) [inline test/test_optimization_pass.cpp:658]: Is this line needed?

**Review by zasdfgbnm — APPROVED** [review]

**Comment by protonu** [comment]
!build

**Pull request merged**